src/init/setup/SetupHttpHandler.ts

```diff
@@ -0,0 +1,215 @@
+ *
+ * GET requests will return the view template which should contain the setup information for the user.
+ * POST requests will run an initializer and/or perform a registration step, both optional.
+ * After successfully completing a POST request this handler will disable itself and become unreachable.
+ * All other methods will be rejected.
+ */
+export class SetupHttpHandler extends BaseHttpHandler {
+  protected readonly logger = getLoggerFor(this);
+
+  private readonly registrationManager?: RegistrationManager;
+  private readonly initializer?: Initializer;
+  private readonly converter: RepresentationConverter;
+  private readonly storage: KeyValueStorage<string, boolean>;
+  private readonly viewTemplate: string;
+  private readonly responseTemplate: string;
+
+  private finished: boolean;
+
+  public constructor(args: SetupHttpHandlerArgs) {
+    super(args);
+    this.finished = false;
+
+    this.registrationManager = args.registrationManager;
+    this.initializer = args.initializer;
+    this.converter = args.converter;
+    this.storage = args.storage;
+    this.viewTemplate = args.viewTemplate;
+    this.responseTemplate = args.responseTemplate;
+  }
+
+  public async canHandle(): Promise<void> {
+    // Check the storage to see if the boolean was set to false due to server restart
+    if (!this.finished) {
+      this.finished = Boolean(await this.storage.get(STORAGE_KEY));
+    }
+
+    if (this.finished) {
+      throw new MethodNotAllowedHttpError('Setup is finished.');
+    }
+  }
+
+  public async handleOperation(operation: Operation): Promise<ResponseDescription> {
+    let json: Record<string, any>;
+    let template: string;
+    let success = false;
+    let statusCode = 200;
+    try {
+      ({ json, template } = await this.getJsonResult(operation));
+      success = true;
+    } catch (err: unknown) {
+      // We want to show the errors on the original page in case of HTML interactions, so we can't just throw them here
+      const error = HttpError.isInstance(err) ? err : new InternalServerError(createErrorMessage(err));
+      ({ statusCode } = error);
+      this.logger.warn(error.message);
+      const response = await this.errorHandler.handleSafe({ error, preferences: { type: { [APPLICATION_JSON]: 1 }}});
+      json = await readJsonStream(response.data!);
+      template = this.viewTemplate;
+    }
+
+    // Convert the response JSON to the required format
+    const representation = new BasicRepresentation(JSON.stringify(json), operation.target, APPLICATION_JSON);
+    addTemplateMetadata(representation.metadata, template, TEXT_HTML);
+    const result = await this.converter.handleSafe(
+      { representation, identifier: operation.target, preferences: operation.preferences },
+    );
+
+    // Make sure this setup handler is never used again after a successful POST request
+    if (success && operation.method === 'POST') {
+      this.finished = true;
+      await this.storage.set(STORAGE_KEY, true);
```

Comment:
We might want to version this instead of a boolean.
Future server versions might require different initializations, and the version number could tell us the delta that needs to be completed.